### ISSUE TITLE
Configure npm registry via .npmrc

### ISSUE THREAD

**Issue by Emptymu** [atlassian/changesets#642]
## Affected Packages
`pkg:@changesets/cli`

## Problem

Looks like we can only get the npm registry from `packageJson.publishConfig` or `process.env.npm_config_registry` when running `npm info`. Is there a way to get npm registry from `.npmrc` file?

<img width="748" alt="Screen Shot 2021-09-28 at 9 06 29 PM" src="https://user-images.githubusercontent.com/25358596/135093465-ed8cd04a-00e9-4091-bbd4-d35e91dcc561.png">

## Proposed solution

Read `npm registry` in the following priorities:

1. `packages/my-package/packageJson.publishConfig`
2. `packages/my-package/.npmrc`
3. `root/.npmrc`

**Comment by Andarist** [comment]
The documentation for `npm_config_registry` this can be found here: https://github.com/npm/cli/blob/30a9844c708482fde3f67a7be833f56891216dd4/docs/content/using-npm/config.md#environment-variables

I'm unsure about the `.npmrc` file - could you tell me how your `.npmrc` looks like? Is the exact thing that you are asking for supported by any other tool?

**Comment by Emptymu** [comment]
hi @Andarist Thanks for linking the `npm cli` doc, it get the npm config following the priority as I expected. And my `.npmrc` file is pretty simple:
```
registry=https://my.registry.org
engine-strict=true

```

Correct me if I'm wrong, looks like the issues is this snippet of code? Where `getCorrectRegistry` doesn't respect `.npmrc` file.
https://github.com/atlassian/changesets/blob/8dee9af60247591ffac75f8f0ef73abb37403159/packages/cli/src/commands/publish/npm-utils.ts#L84-L90
<img width="747" alt="Screen Shot 2021-09-29 at 2 20 54 PM" src="https://user-images.githubusercontent.com/25358596/135213981-4d571931-122c-47de-aaf2-0106be574cb6.png">

Some of my packages were returning `npm info 404` when running `changeset publish` because they're missing `packageJson.publishConfig`, it took me a while to figure out why it's behaving like this. Naturally, I'd expect `changeset publish` to read the npm registry from the `.npmrc` file (just like what `npm cli` would do).

**Comment by Andarist** [comment]
> Correct me if I'm wrong, looks like the issues is this snippet of code? Where getCorrectRegistry doesn't respect .npmrc file.

Yes, if there is an issue then it's here.

> Naturally, I'd expect changeset publish to read the npm registry from the .npmrc file (just like what npm cli would do).

It looks like this proves the issue then - since we want to match the npm cli behavior here. Could you though explain the full setup that you have there and how you are utilizing it (what commands do you invoke through the CLI etc)? This would allow me to see the full picture of the situation at hand.

**Comment by Emptymu** [comment]
Sure, I was running `pnpm changeset publish` at the root level of a pnpm monorepo where `changeset` is also installed at the root level.
```
.npmrc
packages/libs
  - lib-a
    - package.json // has publishConfig.registry
  - lib-b
    - package.json // does not have publishConfig.registry 
```

When publishing, it'll return `npm info 404` for `lib-b`.

I think it's because that `npm cli` command line option has the highest priority, so this command will only use whatever `getCorrectRegistry` returns.

https://github.com/atlassian/changesets/blob/8dee9af60247591ffac75f8f0ef73abb37403159/packages/cli/src/commands/publish/npm-utils.ts#L84-L90

And clearly, `getCorrectRegistry` was not trying to read from `.npmrc` files, it falls back to hard coded registries if there is no `packageJson.publishConfig` nor `process.env.npm_config_registry`.

**Comment by Andarist** [comment]
is `npm info` without the `--registry` flag respecting the registry found in your `.npmrc`? I assume that it's not the default npm/yarn registry that we are talking about, right?

**Comment by Emptymu** [comment]
Just tried removing `"--registry", getCorrectRegistry(packageJson)`, and running `pnpm publish` again. Now correctly got the registry specified in my `.npmrc` file.

**Comment by Andarist** [comment]
Thank you for confirming - I'm confident now that we should adjust the behavior here. I wonder if we shouldn't just let the npm's CLI figure out this on their own. In theory, invoking that from a proper directory should just make it work when I think about it. There is an exception for the yarn registry there so maybe we should keep this and reimplement the logic on our own though.

Do you happen to know if the content of the `.npmrc` is ever automatically loaded into env variables? I guess it isn't as otherwise this would just work thanks to us falling back to `process.env.npm_config_registry`. It would be great to get more knowledge about this subject though.

**Comment by hayes** [comment]
Another side effect of the current pattern is that it makes setting the auth token for the registry a little awkward.  Currently you need to specify the auth token via an npmrc file, and the registry via an environment variable.  When you use the --registry flag, npm won't use the `NPM_TOKEN` environment variable, since that is for the default registry, which means you need to specify a registry specific auth token in your npmrc file.  Needing to define configs for your registry in 2 completely different ways is inconvenient, and took a while to figure out.

**Comment by Andarist** [comment]
@hayes this is good intel - could you summarize exactly what happens now and what is required by changeset plus also how this is supposed to work from your PoV? Listing exact values for all the configs involved in those would be much appreciated

**Comment by Andarist** [comment]
So I knew how to configure the authToken and in general how to configure the registry. I think I understand now better how those 2 play out together - but I have some follow-up questions. Just to make sure that I understand this OK.

If I understand correctly - I can configure multiple auth tokens for multiple registries in a single file because keys for those always include the registry. However, this doesn't touch on what registry will be actually used - so if one wants to use a custom one then they should configure that separately. Would that be right?

> `@some-scope:registry`

I have no idea how to use this. When this gets picked up over `registry`?

> a scoped registry .npmrc file for the scope of the published package

Does this mean that if I opt-in into publishing a particular scope into a custom registry then it's sort of all-or-nothing? All packages of a given scope will pick this up? I guess this is similar to configuring `registry` and publishing a list of unscoped packages. And we always have a way to opt-out a package from a scoped registry by using `publishConfig`.

The default registry from `.npmrc` is still picked up for scoped packages if there is no registry specific for this particular scope, right?

> I think npmrc configs are also inherited from parent directories, so logic for loading those configs might need to traverse up the directory tree looking at each .npmrc file encountered (not sure about this part).

I will make sure to look it up in the implementation of npm.

**Comment by Andarist** [comment]
Mostly a note for me: I've done a quick test and it seems that `npm info` doesn't respect `packageJson.publishConfig.registry` - which makes this very inconvenient for us because we can't simply rely on calling `npm info` either way. Maybe we can use `--registry` only when `packageJson.publishConfig.registry` is defined and delegate figuring out the correct registry to npm in all other cases.

> Probably the easiest way to fix this would be to update the code that looks up the correct registry to pass to the --registry flag to check for 4 places

My current thinking is that we should just avoid using `--registry` as much as possible and I will experiment with that. When done right it feels like it really should yield the best results without us reimplementing a whole bunch of stuff that is a little bit tricky to get 100% right from the start

> npm won't use the NPM_TOKEN environment variable, since that is for the default registry

Do you simply refer to the NPM_TOKEN used by our GitHub action since we only set it up in `.npmrc` for the default registry or are you referring to a broader standard? in other words - Is `NPM_TOKEN` used by npm itself anyhow?

@hayes thank you once again for the detailed write up about this issue. This has helped me to understand the complexity involved here much better. I'm currently working on improving the compatibility with publishing using pnpm and yarn - when I land this work I will be looking into fixing this one as it's a very important feature that is not working correctly right now. I have very limited time resources though - so it won't happen over night, please bear with me 😉

**Comment by shunia** [comment]
@Andarist Sorry to bother but I believe this haven't been implemented right?
I have a similar situation where I'm working with a project using `pnpm` rather than `npm`, and when performing `pnpm changeset publish`, it won't respect  the `registry` value in my `.npmrc`, and will always try to publish the package to the default public npm registry.

I wonder if there is any alternative way to get over this for now if I have to stick with `pnpm`?

**Comment by shunia** [comment]
> @Andarist Sorry to bother but I believe this haven't been implemented right? I have a similar situation where I'm working with a project using `pnpm` rather than `npm`, and when performing `pnpm changeset publish`, it won't respect the `registry` value in my `.npmrc`, and will always try to publish the package to the default npm registry.
> 
> I wonder if there is any alternative way to get over this for now if I have to stick with `pnpm`?

For anyone like me looking for a solution, `pnpm` itself actually provided their way of doing this in the [doc](https://pnpm.io/next/using-changesets), and I will simply paste the process here:

- Run `pnpm changeset version`. This will bump the versions of the packages previously specified with `pnpm changeset` (and any dependents of those) and update the changelog files.
- Run `pnpm install`. This will update the lockfile and rebuild packages.
- Commit the changes.
- Run **`pnpm publish -r`**. This command will publish all packages that have bumped versions not yet present in the registry.
```

**Comment by Andarist** [comment]
I believe this should be fixed by https://github.com/changesets/changesets/pull/2064 . We are going to release a new major version of Changesets with this fix soon. If that won't resolve your issue, please open a new one so we could investigate further.

**Comment by zaman66vpn** [comment]
> ## Affected Packages
> `pkg:@changesets/cli`
> 
> ## Problem
> 
> Looks like we can only get the npm registry from `packageJson.publishConfig` or `process.env.npm_config_registry` when running `npm info`. Is there a way to get npm registry from `.npmrc` file?
> 
> <img width="748" alt="Screen Shot 2021-09-28 at 9 06 29 PM" src="https://user-images.githubusercontent.com/25358596/135093465-ed8cd04a-00e9-4091-bbd4-d35e91dcc561.png">
> 
> Also, is `npm_config_registry` safe to rely on? Because it's not documented.
> 
> ## Proposed solution
> 
> Read `npm registry` in the following priorities:
> 
> 1. `packages/my-package/packageJson.publishConfig`
> 2. `packages/my-package/.npmrc`
> 3. `root/.npmrc`
>

**Comment by zaman66vpn** [comment]
> ## Affected Packages
> `pkg:@changesets/cli`
> 
> ## Problem
> 
> Looks like we can only get the npm registry from `packageJson.publishConfig` or `process.env.npm_config_registry` when running `npm info`. Is there a way to get npm registry from `.npmrc` file?
> 
> <img width="748" alt="Screen Shot 2021-09-28 at 9 06 29 PM" src="https://user-images.githubusercontent.com/25358596/135093465-ed8cd04a-00e9-4091-bbd4-d35e91dcc561.png">
> 
> Also, is `npm_config_registry` safe to rely on? Because it's not documented.
> 
> ## Proposed solution
> 
> Read `npm registry` in the following priorities:
> 
> 1. `packages/my-package/packageJson.publishConfig`
> 2. `packages/my-package/.npmrc`
> 3. `root/.npmrc`
>

**Comment by zaman66vpn** [comment]
> ## Affected Packages
> `pkg:@changesets/cli`
> 
> ## Problem
> 
> Looks like we can only get the npm registry from `packageJson.publishConfig` or `process.env.npm_config_registry` when running `npm info`. Is there a way to get npm registry from `.npmrc` file?
> 
> <img width="748" alt="Screen Shot 2021-09-28 at 9 06 29 PM" src="https://user-images.githubusercontent.com/25358596/135093465-ed8cd04a-00e9-4091-bbd4-d35e91dcc561.png">
> 
> Also, is `npm_config_registry` safe to rely on? Because it's not documented.
> 
> ## Proposed solution
> 
> Read `npm registry` in the following priorities:
> 
> 1. `packages/my-package/packageJson.publishConfig`
> 2. `packages/my-package/.npmrc`
> 3. `root/.npmrc`
>
